include/dsn/utility/metrics.h

```diff
@@ -447,4 +479,268 @@ using concurrent_volatile_counter_ptr = counter_ptr<concurrent_long_adder, true>
 template <typename Adder = striped_long_adder>
 using volatile_counter_prototype = metric_prototype_with<counter<Adder, true>>;
 
+// All supported kinds of kth percentiles. User can configure required kth percentiles for
+// each percentile. Only configured kth percentiles will be computed. This can reduce CPU
+// consumption.
+enum class kth_percentile_type : size_t
+{
+    P50,
+    P90,
+    P95,
+    P99,
+    P999,
+    COUNT,
+    INVALID
+};
+
+// Support to load from configuration files for percentiles.
+ENUM_BEGIN(kth_percentile_type, kth_percentile_type::INVALID)
+ENUM_REG(kth_percentile_type::P50)
+ENUM_REG(kth_percentile_type::P90)
+ENUM_REG(kth_percentile_type::P95)
+ENUM_REG(kth_percentile_type::P99)
+ENUM_REG(kth_percentile_type::P999)
+ENUM_END(kth_percentile_type)
+
+const std::vector<double> kKthDecimals = {0.5, 0.9, 0.95, 0.99, 0.999};
+
+inline size_t kth_percentile_to_nth_index(size_t size, size_t kth_index)
+{
+    auto decimal = kKthDecimals[kth_index];
+    // Since the kth percentile is the value that is greater than k percent of the data values after
+    // ranking them (https://people.richland.edu/james/ictcm/2001/descriptive/helpposition.html),
+    // compute the nth index by size * decimal rather than size * decimal - 1.
+    return static_cast<size_t>(size * decimal);
+}
+
+inline size_t kth_percentile_to_nth_index(size_t size, kth_percentile_type type)
+{
+    return kth_percentile_to_nth_index(size, static_cast<size_t>(type));
+}
+
+std::set<kth_percentile_type> get_all_kth_percentile_types()
+{
+    std::set<kth_percentile_type> all_types;
+    for (size_t i = 0; i < static_cast<size_t>(kth_percentile_type::COUNT); ++i) {
+        all_types.insert(static_cast<kth_percentile_type>(i));
+    }
+    return all_types;
+}
+const std::set<kth_percentile_type> kAllKthPercentileTypes = get_all_kth_percentile_types();
```

Comment:
what about unordered_set ？

Reply:
> what about unordered_set ？

Yes, `std::unordered_set` is a choice. The container for `kth_percentiles` is used to stored all the selected kth percentiles. Since elements in `std::set` are not only unique, but also ordered, we can adopt `std::set` firstly.